x/swaprouter/types/msgs_test.go

```diff
@@ -0,0 +1,326 @@
+		}
+
+		return after(properMsg)
+	}
+
+	msg := createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+		// Do nothing
+		return msg
+	})
+
+	require.Equal(t, msg.Route(), types.RouterKey)
+	require.Equal(t, msg.Type(), "swap_exact_amount_out")
+	signers := msg.GetSigners()
+	require.Equal(t, len(signers), 1)
+	require.Equal(t, signers[0].String(), addr1)
+
+	tests := []struct {
+		name       string
+		msg        types.MsgSwapExactAmountOut
+		expectPass bool
+	}{
+		{
+			name: "proper msg",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				// Do nothing
+				return msg
+			}),
+			expectPass: true,
+		},
+		{
+			name: "invalid sender",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.Sender = invalidAddr.String()
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "empty routes",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.Routes = nil
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "empty routes2",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.Routes = []types.SwapAmountOutRoute{}
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "invalid denom",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.Routes[1].TokenInDenom = "1"
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "invalid denom",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.TokenOut.Denom = "1"
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "zero amount token",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.TokenOut.Amount = sdk.NewInt(0)
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "negative amount token",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.TokenOut.Amount = sdk.NewInt(-10)
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "zero amount criteria",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.TokenInMaxAmount = sdk.NewInt(0)
+				return msg
+			}),
+			expectPass: false,
+		},
+		{
+			name: "negative amount criteria",
+			msg: createMsg(func(msg types.MsgSwapExactAmountOut) types.MsgSwapExactAmountOut {
+				msg.TokenInMaxAmount = sdk.NewInt(-10)
+				return msg
+			}),
+			expectPass: false,
+		},
+	}
+
+	for _, test := range tests {
+		if test.expectPass {
+			require.NoError(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		} else {
+			require.Error(t, test.msg.ValidateBasic(), "test: %v", test.name)
+		}
+	}
+}
+
+// Test authz serialize and de-serializes for swaprouter msg.
+func TestAuthzMsg(t *testing.T) {
+
+	// TODO: uncomment when types are registered.
+	t.SkipNow()
```

Comment:
nit: I think you meant remove when types are registered

Reply:
Will fix this comment in follow-up